designs/auto_scaling/VNF Autoscaling.md

```diff
@@ -0,0 +1,91 @@
+Every VNF which is deployed is configured to use X amount of CPU, memory, storage etc. When a scaling request comes up we want to "hot plug" extra resources(CPU, memory, storage) and the appliance automatically detects this and starts using the extra resources. There are some open issues with scaling down and also there is a limit to the expansion of resources etc
+2. Scale the number of VNF's and load balance traffic across these instances
+Increase or decrease the number of VNF's while using the same resources across all the instance. When "Scaling up" request comes in, increase the number of VNF's and vice versa. This model simplifies the deployment but depending on the security appliance boot up time it might be slower to respond to scale up and scale down requests.
+
+This proposal talks about taking the #2 approach.
+## Assignees
+Arvind Nadendla https://github.com/arvindn05
+## Background
+Currently, the security admin creates a deployment specification which provides the intent for deployment.
+
+We have 2 Basic options
+1. Create a VNF deployment tied to the number hosts within a Availability Zone/region/Host Aggregate etc
+Auto scaling is “built-in” when new hosts get added/deleted VNF also get added/deleted (there is still work that needs to be done to enable this in OSC)
+
+2. Static definition of number of VNF’s per host.
+Here we select the hosts where we want VNF’s to be deployed AND we select the number of instances to be deployed per host. <br/>
+Auto-scaling here means we increase the number of instances while keeping the hosts constant. The intent is the security admin is providing a base line from which you would scale up and scale down.
+
+## Constraints and Assumptions
+The Security manager is monitoring the appliance instances and can determine if a particular instance is overloaded.
+
+## Design Changes
+
+### Deployment
+The admin provides the lower threshold implicitly when creating the deployment specification. In addition, we take an option to specify the upper threshold for the number of instance per host. When a scaling up request comes in, we will deploy extra VNF's upto the upper threshold defined by the administrator.
+
+For example, the deployment specifiation can declare that for 2 hosts H1 and H2 we will deploy 1 VNF instance and the upper theshold is set to 3 VNF instances.
+
+When a scaling request comes up, we are expecting to get the [exact instance which is overloaded as part of the request.](#Constraints-and-Assumptions) Assume H1/VNF1 is this instace <br/>
+OSC will deploy H1/VNF2 the first time and an additional H1/VNF3 if another scaling up request comes in. OSC should NOT deploy any additonal VNF's in H2 since that VNF did not initiate a scale request.
+If another scaling request comes in, OSC has exhausted the scale up policy defined and throws an error.<br/>
+The scale down happens the same way, the extra VNF's are undeployed until we are left with the baseline VNF's only. 
```

Comment:
Today there is an intent of OSC dealing with the scale problem (i.e.: finding the least unassigned instance). Does this mean with this feature we will depart from that completely? If we are relying on the manager to control the scale out/down and on the SDN controller to perform the load balance then i guess it would make sense to remove this intent from OSC, if that is the case maybe mention that in the design as well.